File: apps/home-node-lite/core-server/src/brain/contact_matcher.ts
/**
 * Task 5.35 — contact matcher.
 *
 * Detects mentions of known contacts in free text. Given a user's
 * contact list (names + aliases), builds word-boundary regex
 * patterns once at construction + runs them against each piece of
 * captured text.
 *
 * **Why longest-match-first**:
 *   - If the contact list has both `"Emma"` and `"Emma Watson"`, a
 *     naive scan for `"Emma"` first would claim the span before
 *     `"Emma Watson"` got a chance. Longest-first sorting prevents
 *     the shorter alias from pre-empting the more-specific one.
 *   - Aliases like `"my daughter"` (multi-word) also need priority
 *     over single-word names that happen to appear as substrings
 *     of longer phrases.
 *
 * **Span-claim dedup**: once a pattern claims a text span, shorter
 * patterns skip it. Two different contacts with overlapping names
 * (rare but possible) are resolved by longest-match-wins — matches
 * the Python reference.
 *
 * **Case-insensitive matching** because users rarely capitalise
 * contact names consistently in captured notes. The stored
 * `MatchedContact.name` is the canonical contact name (not the text
 * surface) so downstream code gets a stable id.
 *
 * **Minimum name length 2** prevents false-positive scanning for
 * initials like `"I"` or `"A"`.
 *
 * **Pluggable contact source**: the matcher takes an array at
 * construction. Brain rebuilds the matcher on contact-directory
 * change events — no live-update surface here.
 *
 * Source: docs/HOME_NODE_LITE_TASKS.md Phase 5e task 5.35.
 */

export interface ContactEntry {
  /** Display name (required). Shown in UI and used as canonical id. */
  name: string;
  did: string;
  /** Free-form relationship label (spouse, child, friend, …). */
  relationship?: string;
  /** Who "owns" the data about this contact (household/care/financial/external). */
  dataResponsibility?: string;
  /** Alternate names the user calls this contact (e.g. "mom", "Dr J"). */
  aliases?: string[];
}

export interface MatchedContact {
  /** Canonical display name (not the text surface form). */
  name: string;
  did: string;
  relationship: string;
  dataResponsibility: string;
  /** [start, end) character offsets in the scanned text. */
  span: [number, number];
  /** The literal text that matched (preserved for audit / UI highlight). */
  matchedText: string;
}

const MIN_PATTERN_LENGTH = 2;

interface CompiledPattern {
  pattern: RegExp;
  info: {
    name: string;
    did: string;
    relationship: string;
    dataResponsibility: string;
  };
  /** Length of the raw name — used for longest-first sort. */
  length: number;
}

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export class ContactMatcher {
  private readonly patterns: readonly CompiledPattern[];

  constructor(contacts: ReadonlyArray<ContactEntry>) {
    this.patterns = compilePatterns(contacts);
  }

  /**
   * Find every contact mention in `text`. Returns matches ordered by
   * their position in the text. Overlapping matches are resolved
   * longest-first (shorter patterns skip spans already claimed).
   */
  findMentions(text: string): MatchedContact[] {
    if (typeof text !== 'string' || text.length === 0) return [];
    if (this.patterns.length === 0) return [];

    const results: MatchedContact[] = [];
    const claimed: Array<[number, number]> = [];

    for (const { pattern, info } of this.patterns) {
      pattern.lastIndex = 0;
      for (const m of text.matchAll(pattern)) {
        const start = m.index ?? 0;
        const end = start + m[0]!.length;
        if (overlapsClaimed(start, end, claimed)) continue;
        results.push({
          name: info.name,
          did: info.did,
          relationship: info.relationship,
          dataResponsibility: info.dataResponsibility,
          span: [start, end],
          matchedText: m[0]!,
        });
        claimed.push([start, end]);
      }
    }

    results.sort((a, b) => a.span[0] - b.span[0]);
    return results;
  }

  /** Count of compiled patterns — useful for admin UI + tests. */
  patternCount(): number {
    return this.patterns.length;
  }

  /**
   * Return every compiled pattern's source for debugging. Does NOT
   * expose the regex objects themselves to prevent mutation.
   */
  debugPatterns(): Array<{ source: string; did: string; name: string }> {
    return this.patterns.map((p) => ({
      source: p.pattern.source,
      did: p.info.did,
      name: p.info.name,
    }));
  }
}

// ── Internals ──────────────────────────────────────────────────────────

function compilePatterns(
  contacts: ReadonlyArray<ContactEntry>,
): readonly CompiledPattern[] {
  type Entry = {
    text: string;
    info: CompiledPattern['info'];
  };
  const entries: Entry[] = [];

  for (const c of contacts) {
    const canonicalName = typeof c.name === 'string' ? c.name.trim() : '';
    const info: CompiledPattern['info'] = {
      name: canonicalName,
      did: typeof c.did === 'string' ? c.did : '',
      relationship: typeof c.relationship === 'string' ? c.relationship : 'unknown',
      dataResponsibility:
        typeof c.dataResponsibility === 'string'
          ? c.dataResponsibility
          : 'external',
    };
    if (canonicalName.length >= MIN_PATTERN_LENGTH) {
      entries.push({ text: canonicalName, info });
    }
    if (Array.isArray(c.aliases)) {
      for (const alias of c.aliases) {
        if (typeof alias !== 'string') continue;
        const t = alias.trim();
        if (t.length < MIN_PATTERN_LENGTH) continue;
        entries.push({
          text: t,
          // Aliases inherit the canonical name so downstream code
          // always reports a stable id, even when the surface form
          // differs.
          info: { ...info, name: canonicalName || t },
        });
      }
    }
  }

  // Longest-first so "Emma Watson" matches before "Emma".
  entries.sort((a, b) => b.text.length - a.text.length);

  // Dedup: same DID + same lowered pattern text → one regex.
  const seen = new Set<string>();
  const out: CompiledPattern[] = [];
  for (const e of entries) {
    const key = `${e.info.did}::${e.text.toLowerCase()}`;
    if (seen.has(key)) continue;
    seen.add(key);
    out.push({
      pattern: new RegExp(`\\b${escapeRegex(e.text)}\\b`, 'gi'),
      info: e.info,
      length: e.text.length,
    });
  }
  return out;
}

function overlapsClaimed(
  start: number,
  end: number,
  claimed: ReadonlyArray<[number, number]>,
): boolean {
  // Because patterns run longest-first, a later (shorter) match can
  // only overlap an earlier claim by having its start or end inside
  // the claim range. A shorter match cannot strictly contain a
  // longer claim, so no full-containment check is needed.
  for (const [cs, ce] of claimed) {
    if ((cs <= start && start < ce) || (cs < end && end <= ce)) return true;
  }
  return false;
}

File: legacy/python-brain/src/service/contact_matcher.py
"""Contact mention detection in free text.

Matches known contact names AND aliases against text using word-boundary regex.
Both map to the same DID/contact record.
Longest-match-first ordering prevents partial matches.
"""

from __future__ import annotations

import re
from dataclasses import dataclass


@dataclass(frozen=True)
class MatchedContact:
    """A contact whose name was found in the text."""
    name: str                  # display name that matched
    did: str
    relationship: str          # spouse, child, friend, etc.
    data_responsibility: str   # household, care, financial, external
    span: tuple[int, int]      # (start, end) character offsets in original text


class ContactMatcher:
    """Detects mentions of known contacts in free text.

    Builds word-boundary regex patterns from the contact list.
    Longest-match-first ordering prevents "Jo" matching inside "John".
    Minimum name length 2 characters to avoid false positives.

    Parameters
    ----------
    contacts:
        List of contact dicts with at least: name, did, relationship,
        data_responsibility. Typically from Core's GET /v1/contacts.
    """

    def __init__(self, contacts: list[dict]) -> None:
        self._patterns: list[tuple[re.Pattern, dict]] = []

        # Build patterns for names AND aliases, sorted by length (longest first).
        entries: list[tuple[str, dict]] = []
        for c in contacts:
            info = {
                "did": c.get("did", ""),
                "relationship": c.get("relationship", "unknown"),
                "data_responsibility": c.get("data_responsibility", "external"),
            }

            # Primary name.
            name = (c.get("name") or c.get("display_name") or "").strip()
            if len(name) >= 2:
                entries.append((name, {**info, "name": name}))

            # Aliases (multi-alias list from API response).
            for alias in c.get("aliases", []):
                alias = alias.strip()
                if len(alias) >= 2:
                    entries.append((alias, {**info, "name": name or alias}))

        # Sort longest-first so "my daughter" matches before "daughter",
        # and "Emma Watson" matches before "Emma".
        entries.sort(key=lambda e: len(e[0]), reverse=True)

        # Dedup: same DID + same pattern text → keep only once.
        seen: set[tuple[str, str]] = set()
        for text, info in entries:
            key = (info["did"], text.lower())
            if key in seen:
                continue
            seen.add(key)
            pattern = re.compile(
                r"\b" + re.escape(text) + r"\b",
                re.IGNORECASE,
            )
            self._patterns.append((pattern, info))

    def find_mentions(self, text: str) -> list[MatchedContact]:
        """Find all mentioned contacts in text with character positions.

        Returns one MatchedContact per match. If the same contact matches
        multiple times, each occurrence is returned. Overlapping matches
        from different contacts are resolved longest-first (shorter
        patterns skip spans already claimed by longer ones).
        """
        if not text or not self._patterns:
            return []

        results: list[MatchedContact] = []
        claimed: list[tuple[int, int]] = []  # spans already matched

        for pattern, info in self._patterns:
            for m in pattern.finditer(text):
                span = (m.start(), m.end())
                # Skip if this span overlaps with an already-claimed span.
                if any(s <= span[0] < e or s < span[1] <= e for s, e in claimed):
                    continue
                results.append(MatchedContact(
                    name=info["name"],
                    did=info["did"],
                    relationship=info["relationship"],
                    data_responsibility=info["data_responsibility"],
                    span=span,
                ))
                claimed.append(span)

        # Sort by position in text.
        results.sort(key=lambda mc: mc.span[0])
        return results
